tests/internal/test_wrapping.py

```diff
@@ -491,3 +494,283 @@ def f(a, b, c=None):
+    def foo():
+        raise ValueError("foo")
+
+    wc = DummyWrappingContext(foo)
+    wc.wrap()
+
+    with pytest.raises(ValueError):
+        foo()
+
+    assert wc.entered
+    assert wc.return_value is NOTSET
+    assert wc.exited
+
+    _type, exc, _ = wc.exc_info
+    assert _type == ValueError
+    assert exc.args == ("foo",)
+
+
+def test_wrapping_context_exc_on_exit():
+    class BrokenExitWrappingContext(DummyWrappingContext):
+        def __exit__(self, exc_type, exc_value, traceback):
+            super().__exit__(exc_type, exc_value, traceback)
+            raise RuntimeError("broken")
+
+    def foo():
+        raise ValueError("foo")
+
+    wc = BrokenExitWrappingContext(foo)
+    wc.wrap()
+
+    with pytest.raises(RuntimeError):
+        foo()
+
+    assert wc.entered
+    assert wc.return_value is NOTSET
+    assert wc.exited
+
+    _type, exc, _ = wc.exc_info
+    assert _type == ValueError
+    assert exc.args == ("foo",)
+
+
+def test_wrapping_context_priority():
+    class HighPriorityWrappingContext(DummyWrappingContext):
+        def __enter__(self):
+            nonlocal mutated
+
+            mutated = True
+
+            return super().__enter__()
+
+        def __return__(self, value):
+            nonlocal mutated
+
+            assert not mutated
+
+            return super().__return__(value)
+
+    class LowPriorityWrappingContext(DummyWrappingContext):
+        __priority__ = 99
+
+        def __enter__(self):
+            nonlocal mutated
+
+            assert mutated
+
+            return super().__enter__()
+
+        def __return__(self, value):
+            nonlocal mutated
+
+            mutated = False
+
+            return super().__return__(value)
+
+    mutated = False
+
+    def foo():
+        return 42
+
+    hwc = HighPriorityWrappingContext(foo)
+    lwc = LowPriorityWrappingContext(foo)
+
+    # Wrap low first. We want to make sure that hwc is entered first
+    lwc.wrap()
+    hwc.wrap()
+
+    foo()
+
+    assert lwc.entered
+    assert hwc.return_value == 42
+
+
+def test_wrapping_context_recursive():
+    values = []
+
+    class RecursiveWrappingContext(DummyWrappingContext):
+        def __enter__(self):
+            nonlocal values
+            super().__enter__()
+
+            n = self.__frame__.f_locals["n"]
+            self.set("n", n)
+            values.append(n)
+
+            return self
+
+        def __return__(self, value):
+            nonlocal values
+            n = self.__frame__.f_locals["n"]
+            assert self.get("n") == n
+            values.append(n)
+
+            return super().__return__(value)
+
+    def factorial(n):
+        if n == 0:
+            return 1
+        return n * factorial(n - 1)
+
+    wc = RecursiveWrappingContext(factorial)
+    wc.wrap()
+
+    assert factorial(5) == 120
+    assert values == [5, 4, 3, 2, 1, 0, 0, 1, 2, 3, 4, 5]
+
+
+def test_wrapping_context_generator():
+    def foo():
+        yield from range(10)
```

Comment:
We should be able to add support for that if we need to. This is an interesting topic, as a `yield` cannot close the context, so it would have to be treated differently from a return. More interestingly, what happens for generators that never return?

Reply:
I am happy to yield on the yield topic for later, as long as you think it'll be fine to support later.

generators that never return is at least an existing problem, so I don't think we need a solution today.

designing this context wrapper so people do the right thing in those cases will be great, but not sure we can really detect/determine how best to approach that... ?